### ISSUE TITLE
Inappropriate printing to STDOUT

### ISSUE THREAD

**Issue by znewman01** [sigstore/cosign#2510]
Generally, in Cosign we try to make STDOUT have the *output* of the command: the (often machine-readable) result of execution. We put status messages, errors, etc. on STDERR.

This is a common convention for CLI tools and a nice thing to do for scriptability.

However, we're pretty sloppy sometimes.

---------------------------------------------------------------------
 
I'm looking mostly at:

- https://github.com/sigstore/cosign/blob/fa8a799784cafb0779849e69cb939c8dc6132a5e/pkg/cosign/tlog.go#L170
-  https://github.com/sigstore/cosign/blob/381ba64a577505d378b49482ee34d39eeab6ea50/cmd/cosign/cli/pivcli/commands.go#L152
- https://github.com/sigstore/cosign/blob/381ba64a577505d378b49482ee34d39eeab6ea50/cmd/cosign/cli/initialize/init.go#L52
- https://github.com/sigstore/cosign/blob/d720f040bf493861c1ea8280f462c14e705990e0/cmd/cosign/cli/sign/sign.go#L250

But even some of the ones that are actually correct just work by spitting out to STDOUT in the middle of command execution:
 - https://github.com/sigstore/cosign/blob/381ba64a577505d378b49482ee34d39eeab6ea50/cmd/cosign/cli/pivcli/commands.go#L265
- https://github.com/sigstore/cosign/blob/d720f040bf493861c1ea8280f462c14e705990e0/cmd/cosign/cli/tree.go#L152

-------------------------------------------------------------------

My proposal is the following:

1. Have the CLI top-level entry point *print* the output of the command. That is, something like [`SaveCmd`](https://github.com/sigstore/cosign/blob/30bf1c09c6fde849245648ff294725d86094fe28/cmd/cosign/cli/save.go#L50) would be implemented as a function with a return value, and then we'd hook into Cobra to print the output (if we need to stream output, we could use channels or something).
1. Ban (using [Forbidigo](https://github.com/sigstore/cosign/blob/381ba64a577505d378b49482ee34d39eeab6ea50/.golangci.yml#L38)) any calls to `fmt.Print{f,ln}` in the Cosign codebase.
1. Implement a "Cosign output" package. Instead of `fmt.Fprintln(os.Stderr, "warning: foo")` directly, Cosign code would call `Warn()`. Same for any other common output patterns. (That is, calling code can assign *semantics* to its output which helps make sure it goes to the right place).
1. Ban (most) other `fmt.*` calls (we can keep `fmt.Sprint*`).

This has a number of benefits:

- Supporting multiple formats easily and consistently (you could pass `--format=jsonl` or `--format=csv`) without requiring support from every command
- Testability: parsing STDOUT is hard. Right now that's part of the reason that the `cmd/` directory is under-tested.
- Enforce consistent output: of course, this fixes the STDOUT/STDERR thing. But it also starts to unify the output across the whole tool. If we wanted to start putting all STDERR output in red, we could do that in one place easily. If we want to "tee" all output to a log file, we can do that.

**Comment by eddiezane** [comment]
The way we handle this in [kubectl](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/io_options.go) and [gitsign](https://github.com/sigstore/gitsign/blob/main/internal/io/streams.go) is by encapsulating the streams which has the added bonus of being really easy to mock in tests.

**Comment by znewman01** [comment]
> The way we handle this in [kubectl](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/io_options.go) and [gitsign](https://github.com/sigstore/gitsign/blob/main/internal/io/streams.go) is by encapsulating the streams which has the added bonus of being really easy to mock in tests.

Yeah I think this is also a good idea and can be implemented at the same time we do (3) above.

**Comment by krisharyan117** [comment]
@znewman01 Is this issue still open?

**Comment by kaxada** [comment]
@ralphbean is this issue still open, I am looking at contributing to cosign and think this would be a good start with the getting familiar with the project
